### PR TITLE
Feat #7805: Create the new column in all post types (cached by WP Rocket)

### LINKED ISSUE #7805: Create the new column in all post types (cached by WP Rocket)
**Context**

Here we need to create the new column in posts listing page for all post types that we cache in WP Rocket.

We need a dedicated subscriber for posts.

We can get list of post types using this trait:

https://github.com/wp-media/wp-rocket/blob/c3c54101be5a2a18232530b36ac586a458458ae4/inc/Engine/Optimization/ContentTrait.php#L12

Then we will loop over them in the new subscriber's `get_subscribed_events` method to have something like the following:

```
add_filter( 'manage_' . $post_type . '_posts_columns', 'add_rocket_insights_column' );
function add_rocket_insights_column($columns) {
    $columns['rocket_insights'] = __( 'Rocket Insights', 'rocket' );

    return $columns;
}

// Add the data to the custom columns for the book post type:
add_action( 'manage_' . $post_type . '_posts_custom_column' , 'custom_rocket_insights_column', 10, 2 );
function custom_rocket_insights_column( $column, $post_id ) {
    switch ( $column ) {
        case 'rocket_insights' :
            // Here we will get the url for this post using get_permalink and call the render method which will load the new view dedicated for this layout with any dummy content
            break;

    }
}
```

Then we need to register this new subscriber.

**Expected behavior**

We will have a new column in the public post admin listing page for Rocket Insights with dummy content.

**Acceptance Criteria**

- Have a new column in the public post admin listing page for Rocket Insights with dummy content.

Effort [XS]

### PR DESCRIPTION
**Pull request by Miraeld**
# Description

Fixes #7805 
Display a new column with a placeholder text within the page/post lists.

## Type of change

- [x] New feature (non-breaking change which adds functionality).

### What was tested

Opening the page/post lists pages.
Screenshot 2025-10-17 at 13 34 12
Screenshot 2025-10-17 at 13 34 01

### How to test
Opening the page/post lists pages shows the new column.

### Documentation

This pull request introduces the new "Rocket Insights" column to the post listing pages in the WordPress admin, laying the groundwork for displaying performance insights for each post or page. The implementation includes code changes to add and render the column, dependency injection for rendering, and comprehensive test coverage to ensure correct behavior.

**Key changes:**

**Feature: Rocket Insights Column in Post Listings**
- Added a "Rocket Insights" column to the admin post and page listing tables for all public post types, with logic to exclude types via filter. The column currently displays a placeholder ("Coming Soon") and is ready for future enhancements. [[1]](diffhunk://#diff-df31429d858d608070b6d48d45b0e5a521e2029d09d9f065ff5d78241ff8331bL22-R88) [[2]](diffhunk://#diff-df31429d858d608070b6d48d45b0e5a521e2029d09d9f065ff5d78241ff8331bL77-R183) [[3]](diffhunk://#diff-761ac2485074fb0688a53bacb441cf9fb17d78a4aa5351ab9c4b262ed0318079R1-R18)

**Dependency Management**
- Refactored the `Subscriber` class to inject a `Render` instance via the constructor, and updated the service provider to supply this dependency. [[1]](diffhunk://#diff-df31429d858d608070b6d48d45b0e5a521e2029d09d9f065ff5d78241ff8331bR6) [[2]](diffhunk://#diff-df31429d858d608070b6d48d45b0e5a521e2029d09d9f065ff5d78241ff8331bR15-R31) [[3]](diffhunk://#diff-304c3d096ea5ce72d2acc923ce3f7f69fd91c364b7b9853aedc4cb1033b2e74fL202-R203)

**Rendering Logic**
- Implemented `render_rocket_insights_column` in the `Render` class to output the partial view for the column content.

**Testing**
- Added integration and fixture tests to verify that the column is added, rendered correctly, and handles edge cases (such as missing permalinks or non-target columns). [[1]](diffhunk://#diff-2c1affba7b4d0af13a53e27f6d7092823ce33927586e3dd2bd2bccd30ae574c9R1-R30) [[2]](diffhunk://#diff-3894c8c9dc18e45bd0ea4ce2a1ed5252e27d183251e6e369ff02838b620518fcR1-R61) [[3]](diffhunk://#diff-f09e1c607506b81970cffd130d69053ab2387938d5b6d8390d09ae73189d87a1R1-R50) [[4]](diffhunk://#diff-f8255b08e9d3ea2e41c3f168b19a98cddbdbf27987a1bf4b9dda2c0551cfb0baR1-R51)
# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

### CODE REVIEW

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 1]
```diff
@@ -19,9 +37,55 @@ class Subscriber implements Subscriber_Interface {
...
+		 *
+		 * @return array
+		 */
+		$excluded_post_types = (array) wpm_apply_filters_typed(
+			'array',
+			'rocket_insights_excluded_post_types',
+			[]
+		);
+
+		return array_diff( $post_types, $excluded_post_types );
```
[the remark this answers is not in the document]
Comment (remyperona) [reply inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php:92]: Valid suggestion imo, you could store in a static property

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 2]
```diff
@@ -19,9 +37,55 @@ class Subscriber implements Subscriber_Interface {
 	 * @return array
 	 */
 	public static function get_subscribed_events(): array {
-		return [
+		$post_types = self::get_public_post_types();
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php]: I see @wordpressfan suggested to use the contentTrait [here](https://github.com/wp-media/wp-rocket/issues/7805#issue-3520613326). I think that would have helped not to re-invent the wheel.

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 3]
```diff
@@ -19,9 +37,55 @@ class Subscriber implements Subscriber_Interface {
...
 			'admin_enqueue_scripts' => 'enqueue_post_listing_assets',
 		];
+
+		// Register column hooks for each post type.
+		foreach ( $post_types as $post_type ) {
+			$events[ "manage_{$post_type}_posts_columns" ]       = 'add_rocket_insights_column';
+			$events[ "manage_{$post_type}_posts_custom_column" ] = [ 'render_rocket_insights_column', 10, 2 ];
+		}
+
+		return $events;
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php:62]: WDYT about keeping the `get_subscribed_events` method clean by abstracting the  logic for getting the post types & the loop to a different method which will return only the dynamic events and then in  `get_subscribed_events`, just merge returned arrays?

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 4]
```diff
@@ -19,9 +37,55 @@ class Subscriber implements Subscriber_Interface {
...
+		 * Filters the post types excluded from Rocket Insights on post listing pages.
+		 *
+		 * @since 3.20.1
+		 *
+		 * @param array $excluded_post_types An array of post type names.
+		 *
+		 * @return array
+		 */
+		$excluded_post_types = (array) wpm_apply_filters_typed(
+			'array',
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php]:
```suggestion
			'string[]',
```

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 5]
```diff
@@ -19,11 +40,32 @@ class Subscriber implements Subscriber_Interface {
...
+	 * Build dynamic events for post listing pages based on public post types.
+	 *
+	 * @since 3.20.1
+	 *
+	 * @return array<string, string|array> Associative array of hook => callback(s).
+	 */
+	private static function get_post_listing_events(): array {
+		$events     = [];
+		$post_types = self::get_public_post_type_slugs();
+		// error_log(var_export($post_types, true));
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php]:
```suggestion
```

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

**Comment by jeawhanlee** [comment]
Automated tests and checks are failing.

**Review by jeawhanlee — APPROVED** [review]

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 6]
```diff
@@ -3,14 +3,35 @@
...
 use WP_Rocket\Event_Management\Subscriber_Interface;
+use WP_Rocket\Engine\Optimization\ContentTrait;
 
 /**
  * Subscriber for enqueuing Rocket Insights assets on post listing pages
  *
  * @since 3.20.1
  */
 class Subscriber implements Subscriber_Interface {
+	use ContentTrait;
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php]: No method is used from this ContentTrait, is this still useful?

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 7]
```diff
@@ -19,9 +40,58 @@ class Subscriber implements Subscriber_Interface {
...
+	/**
+	 * Gets the list of public post types that WP Rocket caches.
+	 *
+	 * This is a static helper for get_subscribed_events() since it doesn't require instance state.
+	 *
+	 * @since 3.20.1
+	 *
+	 * @return array
+	 */
+	private static function get_public_post_type_slugs(): array {
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php:74]: This method is also defined in the ContentTrait with a different access modifier.

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php [hunk 8]
```diff
@@ -73,45 +143,46 @@ private function should_enqueue_assets(): bool {
 			return false;
 		}
 
-		// Get the list of public post types that WP Rocket caches.
-		$cached_post_types = $this->get_cached_post_types();
+		$post_type_slugs = self::get_public_post_type_slugs();
```
Comment (jeawhanlee) [inline inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php:144]: In previous commits, we removed this method in favor of `get_public_post_types` from ContentTrait. Why did we bring this back?

File: inc/Engine/Optimization/ContentTrait.php
Comment (jeawhanlee) [inline inc/Engine/Optimization/ContentTrait.php:1]: If we we don't use the changes here, can we revert

**Review by jeawhanlee — DISMISSED** [review]

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php (hunk 6 above)
Comment (Miraeld) [reply inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php]: Nope I removed

File: inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php (hunk 8 above)
Comment (Miraeld) [reply inc/Engine/Admin/RocketInsights/PostListing/Subscriber.php:144]: https://group-onecom.slack.com/archives/C08EFCYRQ2X/p1761118226988059

File: inc/Engine/Optimization/ContentTrait.php
Comment (Miraeld) [reply inc/Engine/Optimization/ContentTrait.php:1]: Reverted

**Review by jeawhanlee — APPROVED** [review]

**Comment by jeawhanlee** [comment]
@Miraeld Great job, can you please port the recent commits to this [PR](https://github.com/wp-media/wp-rocket/pull/7816) as well

**Comment by jeawhanlee** [comment]
Testing this.

**Comment by jeawhanlee** [comment]
I have tested this and it's working as expected. However, we display this column for posts or pages that have "Never cache this page" option selected. @piotrbak Do we need to cater for this?
